### PR TITLE
Send message only once per nick

### LINKED ISSUE #209: Repeating "you have messages"
* nick (....) has joined
\<begiak\> nick: You have messages. Say something, and I'll read them out.
\<begiak\> nick: You have messages. Say something, and I'll read them out.
\<begiak\> nick: You have messages. Say something, and I'll read them out.

### PR DESCRIPTION
**Pull request by Androbin**
Fixes #209

### CODE REVIEW

**Comment by Androbin** [comment]
@sushain97 Friendly ping.

**Comment by sushain97** [comment]
Looks fine to me. Could you write a test?

**Comment by Androbin** [comment]
I was not quite sure about how to construct the `input` argument.
I hope what I took from another test works here as well.

**Comment by Androbin** [comment]
The test fall into the same infinite loop as all other tests do.
It would be kinda out of scope to debug and all of them.

File: test/test_tell.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class BotTest(unittest.TestCase):
+
+    @patch('threading.RLock')
+    @patch('asynchat.async_chat')
+    def setUp(self, mock_async, mock_thread):
+        self.nick = 'foo'
+        self.name = 'Phenny'
+        self.bot = irc.Bot(self.nick, self.name, '#phenny')
```
Comment (shardulc) [inline test/test_tell.py]: Do you really need to instantiate the entire bot? I don't think any of the other tests do this. Just a mock for `self.input` etc. is needed. For example, see [this test](https://github.com/goavki/phenny/blob/master/modules/test/test_remind.py).

**Comment by shardulc** [comment]
Could you explain a little about this infinite loop?

File: test/test_tell.py (hunk 1 above)
Comment (Androbin) [reply test/test_tell.py]: Sorry, I only looked at `test` but not `modules/test`.
Let me try again.

**Comment by Androbin** [comment]
> Could you explain a little about this infinite loop? I don't think I understand what you're referring to.

https://travis-ci.org/goavki/phenny/jobs/320719229#L4674

**Comment by shardulc** [comment]
Interestingly, the infinite loop is not present in the build log for Python 3.6. @Androbin I think you can look into this a little more deeply, at least pointing out what is happening if you can't fix it.

**Comment by Androbin** [comment]
How should `input` look like for a `JOIN` event?

**Comment by shardulc** [comment]
@Androbin I'm not sure. The best way to find out is to see what happens for a live running phenny (`print()`-debugging :P).

File: modules/test/test_tell.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+from modules import tell
+
+
+class TestTell(unittest.TestCase):
+
+    def setUp(self):
+        self.phenny = MagicMock()
+        self.input = MagicMock()
+        self.phenny.nick = 'phenny'
+        self.phenny.config.host = 'test-phenny.example.com'
```
Comment (shardulc) [inline modules/test/test_tell.py]: You don't need this line and the three lines below. They're `remind.py`-specifc.

**Comment by Androbin** [comment]
AFAICT, only the fields `sender` and `nick` should get used, as this is not triggered by user message.

File: modules/test/test_tell.py (hunk 2 above)
Comment (Androbin) [reply modules/test/test_tell.py]: Sorry, I missed that.

**Comment by shardulc** [comment]
@Androbin Sure. Make sure you try out the tests and see the events on your local copy. Also, please don't force-push your commits. It's fine if you have a lot of them.

**Comment by Androbin** [comment]
How do I add aliases from within the test?

**Comment by shardulc** [comment]
Use [the alias method](https://github.com/goavki/phenny/blob/master/modules/tell.py#L53) or any helper methods as you see fit.

**Comment by Androbin** [comment]
Found that myself, I just don't know how to construct the `raw` argument.

**Comment by shardulc** [comment]
It has to be a string matching [this regex](https://github.com/goavki/phenny/blob/master/modules/tell.py#L93). `raw.group(1)` would return the first matched group, etc. Or, as in some other tests, you can mock the `raw.group` method to return strings as you want.

File: modules/test/test_tell.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+from modules import tell
+
+
+class TestTell(unittest.TestCase):
+
+    def setUp(self):
+        self.phenny = MagicMock()
+        self.input = MagicMock()
+        self.phenny.nick = 'phenny'
+        self.phenny.alias_filename = '/home/travis/.phenny/aliases.db'
```
Comment (shardulc) [inline modules/test/test_tell.py]: Now it is not possible to run the test locally... doesn't `~` work? Take a look at the other module tests handle this.

Comment (Androbin) [reply modules/test/test_tell.py]: None of the other tests call either `alias` or `aliasPairMerge` (which calls `dumpAliases`)

Comment (shardulc) [reply modules/test/test_tell.py]: What about calling `dumpAliases` before trying to read the alias database? Even when changed to `~/.phenny/...` I cannot run the test locally.

Comment (Androbin) [reply modules/test/test_tell.py]: And no matter what I try, if gives `FileNotFoundError` even though it should create the file if it doesn't exist. I suspected that it was due to `~` but I can put if back if needed.

Comment (Androbin) [reply modules/test/test_tell.py]: I call `alias` followed by `aliasPairMerge` which calls `dumpAliases` which is why I set this value in the first place.

Comment (shardulc) [reply modules/test/test_tell.py]: `open` with `w` should create the file, is it not working?

Comment (Androbin) [reply modules/test/test_tell.py]: Doesn't seem like it:
https://travis-ci.org/goavki/phenny/jobs/320768872#L855

Comment (Androbin) [reply modules/test/test_tell.py]: What do you suggest?

**Comment by shardulc** [comment]
@Androbin [This test](https://github.com/goavki/phenny/blob/master/modules/test/test_head.py) also relies on a created database, try to do the same thing it does.

**Comment by Androbin** [comment]
Figured it out: `open(fn, 'w')` creates files but not parent directories.

**Comment by shardulc** [comment]
@Androbin Again, please _do not_ force push and change history.

File: modules/test/test_tell.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+class TestTell(unittest.TestCase):
+
+    def setUp(self):
+        self.phenny = MagicMock()
+        self.input = MagicMock()
+        self.phenny.nick = 'phenny'
+
+        alias_file = os.path.expanduser('~/.phenny/aliases.db')
+        os.makedirs(os.path.dirname(alias_file))
```
Comment (shardulc) [inline modules/test/test_tell.py]: `os.makedirs(os.path.dirname(alias_file), exist_ok=True)`

Comment (Androbin) [reply modules/test/test_tell.py]: Good catch

**Comment by Androbin** [comment]
@shardulc Look at the debug logs. Clearly, one of `Testsworth, tester, testing, testmaster` should be in `tester, testing, testmaster`, but somehow it evaluates to `False`.

File: modules/tell.py [hunk 5]
```diff
@@ -266,9 +266,13 @@ def message(phenny, input):
 def messageAlert(phenny, input):
     aliases = aliasGroupFor(input.nick)
     remkeys = map(str.lower, phenny.reminders.keys())
```
Comment (shardulc) [inline modules/tell.py]: Make it `list(map(...))`. A simple `map` object is exhausted after one iteration, the `list` will be stored.

Comment (Androbin) [reply modules/tell.py]: Why has this worked with the old for loop, then?

Comment (sushain97) [reply modules/tell.py]: It probably didn't fully work, lol. (this is why we should have added tests)

**Comment by Androbin** [comment]
Although `phenny.say` seems to be called correctly now, the test says it doesn't get called.

**Comment by shardulc** [comment]
Are you checking `phenny.say` or `phenny.msg`?

**Comment by Androbin** [comment]
Currently, I am testing via `phenny.msg` but that shouldn't be a problem?

**Comment by Androbin** [comment]
Okay, it works when testing `phenny.say`, great!
But I still don't know what I messed up when testing `phenny.msg`?

**Comment by shardulc** [comment]
`phenny.msg` and `phenny.say` are two different methods, and the testing framework checks for calls to methods and not the end result.

**Comment by Androbin** [comment]
@shardulc I am aware of that, I passed the arguments that `phenny.msg` was supposed to be eventually called with.

**Comment by sushain97** [comment]
The mock library might not be registering methods that aren't directly called by the tests as being called.
